File: src/modules/approval/services/approval-workflow.service.ts
import { Injectable } from '@nestjs/common';
import { EntityManager } from 'typeorm';
import { Employee } from '../../employee/entities/employee.entity';
import { EquipmentRequest } from '../../request/entities/equipment-request.entity';
import { ApprovalStep } from '../entities/approval-step.entity';
import { ApprovalRole } from '../enums/approval-role.enum';
import { ApprovalStepStatus } from '../enums/approval-step-status.enum';

@Injectable()
export class ApprovalWorkflowService {
  async createInitialManagerApprovalStep(
    manager: EntityManager,
    request: EquipmentRequest,
    directManager: Employee,
  ): Promise<ApprovalStep> {
    const stepRepo = manager.getRepository(ApprovalStep);

    const managerStep = stepRepo.create({
      request,
      level: 1,
      approver: directManager,
      approverRole: ApprovalRole.DIRECT_MANAGER,
      status: ApprovalStepStatus.PENDING,
    });

    return stepRepo.save(managerStep);
  }
}

File: src/EquipmentApi/Services/ApprovalWorkflowService.cs
using EquipmentApi.Data;
using EquipmentApi.Entities;
using EquipmentApi.Enums;
using Microsoft.EntityFrameworkCore;

namespace EquipmentApi.Services;

public class ApprovalWorkflowService
{
    public async Task<ApprovalStep> CreateInitialManagerApprovalStepAsync(
        AppDbContext db,
        EquipmentRequest request,
        Employee directManager,
        CancellationToken cancellationToken = default)
    {
        var managerStep = new ApprovalStep
        {
            RequestId = request.Id,
            Level = 1,
            ApproverId = directManager.Id,
            ApproverRole = ApprovalRole.DirectManager,
            Status = ApprovalStepStatus.Pending,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow,
        };

        db.ApprovalSteps.Add(managerStep);
        await db.SaveChangesAsync(cancellationToken);
        managerStep.Approver = directManager;
        return managerStep;
    }
}
